plat/common/include/arm/arm64/paging.h

```diff
@@ -0,0 +1,381 @@
+		pte |= PTE_ATTR_IDX(DEVICE_GRE);
+		break;
+	default:
+		UK_ASSERT(0 && "Invalid memory type\n");
+	}
+
+	return pte;
+}
+
+static inline __pte_t
+pgarch_pte_change_attr(__pte_t pte, unsigned long new_attr,
+		       unsigned int level __unused)
+{
+	(void)pte;
+	(void)new_attr;
+
+	UK_CRASH("%s: Not implemented", __func__);
+
+	return 0;
+}
+
+static inline unsigned long
+pgarch_attr_from_pte(__pte_t pte, unsigned int level __unused)
+{
+	unsigned long attr = PAGE_ATTR_PROT_READ;
+
+	if (!(pte & PTE_ATTR_AP(PTE_ATTR_AP_RO)))
+		attr |= PAGE_ATTR_PROT_WRITE;
+
+	if (!(pte & PTE_ATTR_PXN))
+		attr |= PAGE_ATTR_PROT_EXEC;
+
+	switch (pte & PTE_ATTR_SH_MASK) {
+	case (PTE_ATTR_SH(PTE_ATTR_SH_NS)):
+		attr |= PAGE_ATTR_SHAREABLE_NS;
+		break;
+	case (PTE_ATTR_SH(PTE_ATTR_SH_IS)):
+		attr |= PAGE_ATTR_SHAREABLE_IS;
+		break;
+	case (PTE_ATTR_SH(PTE_ATTR_SH_OS)):
+		attr |= PAGE_ATTR_SHAREABLE_OS;
+		break;
+	default:
+		UK_ASSERT(0 && "Invalid shareability type\n");
+	};
+
+	switch (pte & PTE_ATTR_IDX_MASK) {
+	case PTE_ATTR_IDX(NORMAL_WB):
+		attr |= PAGE_ATTR_TYPE_NORMAL_WB;
+		break;
+	case PTE_ATTR_IDX(NORMAL_WT):
+		attr |= PAGE_ATTR_TYPE_NORMAL_WT;
+		break;
+	case PTE_ATTR_IDX(NORMAL_NC):
+		attr |= PAGE_ATTR_TYPE_NORMAL_NC;
+		break;
+	case PTE_ATTR_IDX(DEVICE_nGnRnE):
+		attr |= PAGE_ATTR_TYPE_DEVICE_nGnRnE;
+		break;
+	case PTE_ATTR_IDX(DEVICE_nGnRE):
+		attr |= PAGE_ATTR_TYPE_DEVICE_nGnRE;
+		break;
+	case PTE_ATTR_IDX(DEVICE_GRE):
+		attr |= PAGE_ATTR_TYPE_DEVICE_GRE;
+		break;
+	default:
+		UK_ASSERT(0 && "Invalid memory type\n");
+	}
+
+	return attr;
+}
+
+/*
+ * Page tables
+ */
+static inline __vaddr_t
+pgarch_pt_pte_to_vaddr(struct uk_pagetable *pt __unused, __pte_t pte,
+		       unsigned int level __maybe_unused)
+{
+	(void)pt;
+
+	return arm64_directmap_paddr_to_vaddr(PT_Lx_PTE_PADDR(pte, level));
+}
+
+/* Create a table descriptor */
+static inline __pte_t
+pgarch_pt_pte_create(struct uk_pagetable *pt __unused, __paddr_t pt_paddr,
+		     unsigned int level __unused, __pte_t template,
+		     unsigned int template_level __unused)
+{
+	(void)template;
+
+	__pte_t pt_pte;
+
+	UK_ASSERT(PAGE_ALIGNED(pt_paddr));
+
+	pt_pte = pt_paddr & PTE_Lx_TABLE_PADDR_MASK;
+
+	pt_pte |= PTE_TYPE_TABLE;
+
+	return pt_pte;
+}
+
+static inline __vaddr_t
+pgarch_pt_map(struct uk_pagetable *pt __unused, __paddr_t pt_paddr,
+	      unsigned int level __unused)
+{
+	return arm64_directmap_paddr_to_vaddr(pt_paddr);
+}
+
+static inline __paddr_t
+pgarch_pt_unmap(struct uk_pagetable *pt __unused, __vaddr_t pt_vaddr,
+		unsigned int level __unused)
+{
+	return arm64_directmap_vaddr_to_paddr(pt_vaddr);
+}
+
+static inline int pgarch_init(void)
+{
+	/* Sanity checks to make sure that the PE supports the minimum
+	 * requirements and the platform has set up a sane environment.
+	 * This is expected to be useful during bring-up, so keep it
+	 * conditional to UKDEBUG, to improve boot performance.
+	 */
+#ifdef CONFIG_LIBUKDEBUG
```

Comment:
Actually, what makes more sense here is the following because `ukdebug` is effectively part of every build. Ensure that you match the `endif`.

```suggestion
#ifdef CONFIG_UKDEBUG
```